tests/test_pickling.py

```diff
@@ -45,3 +45,35 @@ def test_enum_pickle():
 
     data = pickle.dumps(e.EOne, 2)
     assert e.EOne == pickle.loads(data)
+
+
+#
+# exercise_trampoline
+#
+class SimplePyDerived(m.SimpleBase):
+    pass
+
+
+def test_roundtrip_simple_py_derived():
+    p = SimplePyDerived()
+    p.num = 202
+    p.stored_in_dict = 303
+    data = pickle.dumps(p, pickle.HIGHEST_PROTOCOL)
+    p2 = pickle.loads(data)
+    assert isinstance(p2, SimplePyDerived)
+    assert p2.num == 202
+    assert p2.stored_in_dict == 303
+
+
+def test_roundtrip_simple_cpp_derived():
+    p = m.make_SimpleCppDerivedAsBase()
+    p.num = 404
+    if not env.PYPY:
+        # To ensure that this unit test is not accidentally invalidated.
+        with pytest.raises(AttributeError):
+            # Mimics the `setstate` C++ implementation.
+            setattr(p, "__dict__", {})  # noqa: B010
+    data = pickle.dumps(p, pickle.HIGHEST_PROTOCOL)
+    p2 = pickle.loads(data)
+    assert isinstance(p2, m.SimpleBase)
```

Comment:
This test now makes me concerned about object slicing: https://en.wikipedia.org/wiki/Object_slicing
As I read this, this object is now *sliced*; it was originally `SimpleCppDerived`, but Python only knows it as `SimpleBase`, therefore the object (and its vtable) will get sliced down to `SimpleBase`.

Ideally, there's a fix, either:
(a) If you pickle a `typeid` that isn't explicitly registered with pybind11, fail fast (otherwise, pybind11 doesn't know how to reconsistitute)
(b) Somehow, magically, teach pybind11 how to pickle+unpickle lower type hierarchies (I dunno how to do this without dark magic)

That being said, I'm totally fine with that fix being delegated to an issue+TODO (if it doesn't already exist), as this PR is *definitely* a strict improvement.

(Is this already a documented defect in `pybind11`?)

Reply:
Filed #3062

Can you refer to that, and perhaps steal some of the testing expectations I added in my branch?

Reply:
> Filed #3062
> 
> Can you refer to that, and perhaps steal some of the testing expectations I added in my branch?

Done, although I used a `dynamic_cast` instead (before I realized there was test code under #3062).

Based on @elkhrt's [comment](https://github.com/pybind/pybind11/issues/3062#issuecomment-871499868) I believe the situation that is helped by the change here is a valid use case.